File: solutions/cpp/solution1/src/video.cpp
#include "video.h"

#include <iostream>
#include <utility>
#include <vector>

Video::Video(std::string&& title, std::string&& videoId,
             std::vector<std::string>&& tags) :
  mTitle(std::move(title)),
  mVideoId(std::move(videoId)),
  mTags(std::move(tags)) {
}

const std::string& Video::getTitle() const { return mTitle; }

const std::string& Video::getVideoId() const { return mVideoId; }

const std::vector<std::string>& Video::getTags() const { return mTags; }

const std::string Video::getTagsAsStr() const {
    std::string tag_str = "";
    for (const auto& tag : mTags) {
        tag_str = tag_str + tag + " ";
    }
    tag_str = tag_str.substr(0, tag_str.size() - 1);
    return tag_str;
}

File: solutions/java/ricdotnet/src/main/java/com/google/Video.java
package com.google;

import java.util.Collections;
import java.util.List;

/** A class used to represent a video. */

/**
 * USER NOTES:
 * Implemented Comparable and compareTo in order to lexicographically order the videos by title.
 */
class Video implements Comparable<Video> {

  private final String title;
  private final String videoId;
  private final List<String> tags;
  private Boolean isFlagged = false;
  private String flaggedReason = null;

  Video(String title, String videoId, List<String> tags) {
    this.title = title;
    this.videoId = videoId;
    this.tags = Collections.unmodifiableList(tags);
  }

  /** Returns the title of the video. */
  String getTitle() {
    return title;
  }

  /** Returns the video id of the video. */
  String getVideoId() {
    return videoId;
  }

  /** Returns a readonly collection of the tags of the video. */
  List<String> getTags() {
    return tags;
  }

  /** Sets the isFlagged state of the video **/
  void setIsFlagged(Boolean isFlagged) {
    this.isFlagged = isFlagged;
  }

  /** Returns the isFlagged state of a video **/
  Boolean getIsFlagged() {
    return isFlagged;
  }

  /** Sets the flagging reason for the video **/
  void setFlaggedReason(String flaggedReason) {
    this.flaggedReason = flaggedReason;
  }

  /** Returns the flagging reason for the video **/
  String getFlaggedReason() {
    return flaggedReason;
  }

  @Override
  public int compareTo(Video video) {
//    if(this.getTitle().compareTo(video.getTitle()) > 0) {
//      return 1;
//    } else if(this.getTitle().compareTo(video.getTitle()) < 0) {
//      return -1;
//    } else {
//      return 0;
//    }
    return Integer.compare(this.title.compareTo(video.title), 0);
  }
}
